File: tests/python/test_big_int.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
from __future__ import annotations

import sys

import pytest
import tvm_ffi
import tvm_ffi.cpp
from tvm_ffi.core import TypeSchema

# Distinct words and an interior zero exercise word order and sign padding.
_MULTIWORD_VALUE = (1 << 255) | (0x0123456789ABCDEF << 128) | 0xFEDCBA9876543210


@pytest.fixture(scope="module")
def bigint_module() -> tvm_ffi.Module:
    return tvm_ffi.cpp.load_inline(
        name="test_big_int_transport",
        cpp_sources=r"""
            #include <tvm/ffi/big_int.h>
            #include <tvm/ffi/rvalue_ref.h>

            tvm::ffi::BigInt identity(tvm::ffi::BigInt value) { return value; }
            tvm::ffi::BigInt add(tvm::ffi::BigInt a, tvm::ffi::BigInt b) { return a + b; }
            bool is_int64(tvm::ffi::Any value) {
              return value.type_index() == kTVMFFIInt;
            }
            int64_t int64_only(int64_t value) { return value; }
            tvm::ffi::BigInt invoke(tvm::ffi::Function callback, tvm::ffi::BigInt value) {
              return callback(value).cast<tvm::ffi::BigInt>();
            }
            tvm::ffi::BigInt invoke_rvalue(tvm::ffi::Function callback, tvm::ffi::BigInt value) {
              auto object = tvm::ffi::Any(value).cast<tvm::ffi::ObjectRef>();
              return callback(tvm::ffi::RValueRef(std::move(object))).cast<tvm::ffi::BigInt>();
            }
            int64_t rvalue_ref_count_delta(tvm::ffi::Function callback, tvm::ffi::BigInt value) {
              auto owner = tvm::ffi::Any(value).cast<tvm::ffi::ObjectRef>();
              int64_t before = owner.use_count();
              try {
                auto moved = owner;
                callback(tvm::ffi::RValueRef(std::move(moved)));
              } catch (const tvm::ffi::Error&) {
              }
              return owner.use_count() - before;
            }
            tvm::ffi::Bytes content(tvm::ffi::BigInt value) {
              TVMFFIAny view = tvm::ffi::AnyView(value).CopyToTVMFFIAny();
              TVMFFIByteArray bytes = TVMFFIBigIntGetContentByteArray(&view);
              return tvm::ffi::Bytes(bytes.data, bytes.size);
            }
            tvm::ffi::Any from_content(tvm::ffi::Bytes bytes) {
              TVMFFIByteArray input{bytes.data(), bytes.size()};
              TVMFFIAny result;
              if (TVMFFIBigIntFromByteArray(&input, &result) != 0) {
                throw tvm::ffi::details::MoveFromSafeCallRaised();
              }
              return tvm::ffi::details::AnyUnsafe::MoveTVMFFIAnyRawToAny(result);
            }
        """,
        functions=[
            "identity",
            "add",
            "is_int64",
            "int64_only",
            "invoke",
            "invoke_rvalue",
            "rvalue_ref_count_delta",
            "content",
            "from_content",
        ],
        extra_cflags=["-DTVM_FFI_DLL_EXPORT_INCLUDE_METADATA=1"],
    )


def test_integer_round_trip(bigint_module: tvm_ffi.Module) -> None:
    echo = tvm_ffi.get_global_func("testing.echo")
    values = [0, -1, 0x13579BDF2468ACE0, -0x13579BDF2468ACE0, _MULTIWORD_VALUE, -_MULTIWORD_VALUE]
    for value in values:
        assert echo(value) == value
        result = bigint_module.identity(value)
        assert type(result) is int
        assert result == value
        assert bigint_module.is_int64(value) == (-(1 << 63) <= value < (1 << 63))


def test_integer_int64_boundaries(bigint_module: tvm_ffi.Module) -> None:
    echo = tvm_ffi.get_global_func("testing.echo")
    for value in [-(1 << 63) - 1, -(1 << 63), (1 << 63) - 1, 1 << 63]:
        assert echo(value) == value
        assert bigint_module.identity(value) == value
        fits = -(1 << 63) <= value < (1 << 63)
        assert bigint_module.is_int64(value) == fits
        if fits:
            assert bigint_module.int64_only(value) == value


def test_integer_content_format(bigint_module: tvm_ffi.Module) -> None:
    assert bigint_module.from_content(b"") == 0
    with pytest.raises(ValueError, match="whole 64-bit words"):
        bigint_module.from_content(b"\x00" * 7)
    values = [0, -1, (1 << 63) - 1, -(1 << 63), 1 << 63, -(1 << 63) - 1]
    values.extend([_MULTIWORD_VALUE, -_MULTIWORD_VALUE])
    for value in values:
        signed_bits = (value if value >= 0 else ~value).bit_length() + 1
        word_count = (signed_bits + 63) // 64
        expected = b"".join(
            ((value >> (64 * i)) & ((1 << 64) - 1)).to_bytes(8, sys.byteorder)
            for i in range(word_count)
        )
        content = bigint_module.content(value)
        assert content == expected
        assert bigint_module.from_content(content) == value
        padding = (b"\xff" if value < 0 else b"\x00") * 16
        assert bigint_module.from_content(content + padding) == value


def test_integer_promotion_and_demotion(bigint_module: tvm_ffi.Module) -> None:
    assert bigint_module.add((1 << 63) - 1, 1) == 1 << 63
    assert bigint_module.add(-(1 << 63), -1) == -(1 << 63) - 1
    for value in [1 << 63, -(1 << 63) - 1, _MULTIWORD_VALUE, -_MULTIWORD_VALUE]:
        assert bigint_module.add(value, -value) == 0
        assert bigint_module.is_int64(bigint_module.add(value, 17 - value))
        with pytest.raises(TypeError):
            bigint_module.int64_only(value)


def test_integer_type_schema(bigint_module: tvm_ffi.Module) -> None:
    metadata = bigint_module.get_function_metadata("identity")
    assert metadata is not None
    assert "ffi.BigInt" in metadata["type_schema"]
    assert str(TypeSchema.from_json_str(metadata["type_schema"])) == "Callable[[int], int]"


def test_integer_callbacks(bigint_module: tvm_ffi.Module) -> None:
    value = _MULTIWORD_VALUE
    received = []

    def callback(arg: int) -> int:
        assert type(arg) is int
        received.append(arg)
        return -arg - 17

    expected = -value - 17
    assert bigint_module.invoke(callback, value) == expected
    assert tvm_ffi.convert_func(callback)(value) == expected
    assert received == [value, value]
    assert tvm_ffi.convert_func(lambda: value)() == value


def test_integer_rvalue_callbacks(bigint_module: tvm_ffi.Module) -> None:
    def identity(value: int) -> int:
        assert type(value) is int
        return value

    def failing(value: int) -> int:
        assert type(value) is int
        raise ValueError("callback failed")

    value = _MULTIWORD_VALUE
    assert bigint_module.invoke_rvalue(identity, value) == value
    assert bigint_module.rvalue_ref_count_delta(identity, value) == 0
    with pytest.raises(ValueError, match="callback failed"):
        bigint_module.invoke_rvalue(failing, value)
    assert bigint_module.rvalue_ref_count_delta(failing, value) == 0


def test_integer_containers() -> None:
    values = [_MULTIWORD_VALUE, -_MULTIWORD_VALUE, 0, (1 << 63) - 1]
    echo = tvm_ffi.get_global_func("testing.echo")
    array = echo(values)
    assert list(array) == values
    mapping = echo({value: -value for value in values})
    assert len(mapping) == len(values)
    for value in values:
        assert mapping[value] == -value


def test_numpy_unsigned_integer() -> None:
    np = pytest.importorskip("numpy")
    echo = tvm_ffi.get_global_func("testing.echo")
    assert echo(np.uint64((1 << 64) - 1)) == (1 << 64) - 1

File: tests/cpp/extra/test_big_int.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <gtest/gtest.h>
#include <tvm/ffi/big_int.h>
#include <tvm/ffi/container/array.h>
#include <tvm/ffi/container/map.h>
#include <tvm/ffi/extra/dataclass.h>
#include <tvm/ffi/extra/serialization.h>
#include <tvm/ffi/extra/structural_equal.h>
#include <tvm/ffi/extra/structural_hash.h>
#include <tvm/ffi/extra/structural_mutate.h>

#include <limits>
#include <sstream>
#include <utility>

namespace {
using namespace tvm::ffi;
TEST(BigIntExtra, JSONGraphRoundTrip) {
  BigInt wide = BigInt(1) << 100;
  for (const auto& [value, decimal] :
       {std::pair<BigInt, String>{wide, "1267650600228229401496703205376"},
        {-wide, "-1267650600228229401496703205376"},
        {BigInt(std::numeric_limits<int64_t>::max()) + 1, "9223372036854775808"},
        {BigInt(std::numeric_limits<int64_t>::min()) - 1, "-9223372036854775809"}}) {
    SCOPED_TRACE(decimal);
    json::Object expected{
        {"root_index", 0},
        {"nodes", json::Array{json::Object{{"type", "ffi.BigInt"}, {"data", decimal}}}}};
    auto graph = ToJSONGraph(value);
    EXPECT_TRUE(StructuralEqual()(graph, expected));
    Any restored = FromJSONGraph(json::Parse(json::Stringify(graph)));
    EXPECT_EQ(restored.type_index(), TypeIndex::kTVMFFIBigInt);
    EXPECT_EQ(restored.cast<BigInt>(), value);
  }
  Array<Any> nested{7, Map<String, Any>{{"values", Array<Any>{wide, -wide}}}};
  EXPECT_TRUE(
      StructuralEqual()(FromJSONGraph(json::Parse(json::Stringify(ToJSONGraph(nested)))), nested));
}

TEST(BigIntExtra, JSONGraphCanonicalInt64) {
  for (const auto& [decimal, value] :
       {std::pair<String, int64_t>{"0", 0},
        {"-0", 0},
        {"+0007", 7},
        {"9223372036854775807", std::numeric_limits<int64_t>::max()},
        {"-9223372036854775808", std::numeric_limits<int64_t>::min()}}) {
    SCOPED_TRACE(decimal);
    json::Object graph{
        {"root_index", 0},
        {"nodes", json::Array{json::Object{{"type", "ffi.BigInt"}, {"data", decimal}}}}};
    Any restored = FromJSONGraph(graph);
    EXPECT_EQ(restored.type_index(), TypeIndex::kTVMFFIInt);
    EXPECT_EQ(restored.cast<int64_t>(), value);
    json::Object expected{{"root_index", 0},
                          {"nodes", json::Array{json::Object{{"type", "int"}, {"data", value}}}}};
    EXPECT_TRUE(StructuralEqual()(ToJSONGraph(BigInt(value)), expected));
    EXPECT_EQ(FromJSONGraph(expected).cast<BigInt>(), value);
  }
}

TEST(BigIntExtra, JSONGraphInvalidData) {
  for (const Any& data : Array<Any>{"", "+", "-", "x", "12x", " 1", "1 ", "1.0", String("12\0x", 4),
                                    nullptr, true, 1, 1.0, json::Array{}, json::Object{}}) {
    SCOPED_TRACE(ReprPrint(data));
    json::Object graph{
        {"root_index", 0},
        {"nodes", json::Array{json::Object{{"type", "ffi.BigInt"}, {"data", data}}}}};
    EXPECT_THROW(FromJSONGraph(graph), Error);
  }
}

TEST(BigIntExtra, NumericValueAndImmutableCopy) {
  BigInt a = BigInt(1) << 255;
  BigInt b = BigInt(2) << 254;
  BigInt c = BigInt(1) << 256;
  EXPECT_TRUE(StructuralEqual()(a, b));
  EXPECT_FALSE(StructuralEqual()(a, c));
  EXPECT_EQ(StructuralHash()(a), StructuralHash()(b));
  EXPECT_TRUE(RecursiveEq(a, b));
  EXPECT_FALSE(RecursiveEq(a, c));
  EXPECT_TRUE(RecursiveLt(a, c));
  BigInt next = a + 1;
  EXPECT_FALSE(StructuralEqual()(a, next));
  EXPECT_FALSE(RecursiveEq(a, next));
  EXPECT_TRUE(RecursiveLt(a, next));
  EXPECT_TRUE(RecursiveLt(-c, -a));
  EXPECT_TRUE(RecursiveLt(-a, b));
  EXPECT_TRUE(RecursiveGt(a, -b));
  EXPECT_EQ(RecursiveHash(a), RecursiveHash(b));
  Array<Any> nested{a, -c};
  Array<Any> equal{b, -c};
  EXPECT_TRUE(RecursiveEq(nested, equal));
  EXPECT_EQ(RecursiveHash(nested), RecursiveHash(equal));
  EXPECT_TRUE(DeepCopy(a).same_as(Any(a)));
  EXPECT_TRUE(RecursiveEq(DeepCopy(nested), nested));
  std::ostringstream stream;
  stream << a;
  EXPECT_EQ(ReprPrint(a), stream.str());
}
TEST(BigIntExtra, RecursiveOrderingAcrossRepresentations) {
  BigInt max = std::numeric_limits<int64_t>::max();
  BigInt min = std::numeric_limits<int64_t>::min();
  BigInt above = max + 1;
  BigInt below = min - 1;
  EXPECT_TRUE(RecursiveLt(max, above));
  EXPECT_TRUE(RecursiveGt(above, max));
  EXPECT_FALSE(RecursiveLt(above, max));
  EXPECT_TRUE(RecursiveLt(below, min));
  EXPECT_TRUE(RecursiveGt(min, below));
  EXPECT_FALSE(RecursiveLt(min, below));
  EXPECT_TRUE(RecursiveLt(below, 1));
  EXPECT_TRUE(RecursiveGt(1, below));
  EXPECT_TRUE(RecursiveLt(Array<Any>{7, max}, Array<Any>{7, above}));
  EXPECT_TRUE(RecursiveLt(Array<Any>{7, below}, Array<Any>{7, min}));
  EXPECT_FALSE(RecursiveEq(max, above));
  EXPECT_FALSE(RecursiveEq(below, min));
  EXPECT_THROW(RecursiveLt(1.0, above), Error);
  EXPECT_THROW(RecursiveLt(above, String("1")), Error);
  EXPECT_THROW(RecursiveLt(above, Array<Any>{1}), Error);
}
TEST(BigIntExtra, StructuralMutation) {
  BigInt value = BigInt(1) << 255;
  BigInt alias = value;
  auto never_matches = [](int64_t, StructuralMutatorObj*) -> Expected<Any> { return Any(); };
  using Mutator = StructuralMutateEngine<StructuralMapEngineBase, decltype(never_matches)>;
  StructuralMutator mutator(make_object<Mutator>(never_matches));
  EXPECT_TRUE(mutator->Mutate(AnyView(value)).IsUnchanged());
  EXPECT_TRUE(mutator->MaybeInplaceMutate(AnyView(value)).IsUnchanged());
  EXPECT_TRUE(StructuralMutateExpected(Any(value), never_matches).value().same_as(Any(value)));
  auto map_int = [](int64_t x) -> Expected<Any> { return Any(x + 1); };
  EXPECT_TRUE(StructuralMapExpected<WalkOrder::kPostOrder>(Any(value), map_int)
                  .value()
                  .same_as(Any(value)));
  Array<Any> nested{value, 1};
  auto result =
      StructuralMapExpected<WalkOrder::kPostOrder>(Any(nested), map_int).value().cast<Array<Any>>();
  EXPECT_TRUE(result[0].same_as(Any(alias)));
  EXPECT_EQ(result[1].cast<int64_t>(), 2);
  EXPECT_EQ(value, alias);
}
TEST(BigIntExtra, StructuralMapKeyHashDoesNotDependOnAliases) {
  BigInt a = BigInt(1) << 255;
  BigInt b = BigInt(2) << 254;
  BigInt c = BigInt(4) << 253;
  Array<Any> lhs{a, Map<Any, Any>{{a, 7}}};
  Array<Any> rhs{b, Map<Any, Any>{{c, 7}}};
  EXPECT_TRUE(StructuralEqual()(lhs, rhs));
  EXPECT_EQ(StructuralHash()(lhs), StructuralHash()(rhs));
  Map<Any, Any> first{{a, 7}, {-a, 8}};
  EXPECT_EQ(first[b].cast<int64_t>(), 7);
  Map<Any, Any> second{{-b, 8}, {c, 7}};
  EXPECT_TRUE(StructuralEqual()(first, second));
  EXPECT_EQ(StructuralHash()(first), StructuralHash()(second));
  EXPECT_NE(StructuralHash()(first), StructuralHash()(Map<Any, Any>{{a, 9}, {-a, 8}}));
}
}  // namespace
